### ISSUE TITLE
semver.format function is dropping build data

### ISSUE THREAD

**Issue by justinmchase** [denoland/deno_std#3110]
**Describe the bug**
When formatting a semver object, the build data is not printed even if its on the original semver object.

**Steps to Reproduce**

```ts
import { parse } from "https://deno.land/std@0.172.0/semver/mod.ts";
const v = parse('1.2.3+4')
console.log(v.format()) // 1.2.3, expected 1.2.3+4
```

**Expected behavior**

I expect the build metadata to continue to be printed as part of the version

**Environment**

- deno 1.28.3 (release, x86_64-apple-darwin)
- std version: 0.172.0

### Recommended Function
```ts
function format(semver: SemVer): string {
  const main = `${semver.major}.${semver.minor}.${semver.patch}`;
  const pre = semver.prerelease.join(".");
  const build = semver.build.join(".");
  return [[main, pre].filter(_ => _).join('-'), build].filter(_ => _).join("+")
}
```

**Comment by justinmchase** [comment]
Additionally there is no function to replace the build values. You can alter the pre-release by calling the `.increment` function and providing a new pre-release name but there is no way to increment or even replace the build number with out doing string manipulation and re-parsing.

**Comment by kamilogorek** [comment]
This is exact same implementation as https://www.npmjs.com/package/semver uses, so there's a question whether Deno should be compatible with it, or be "more correct" with some of its methods.
`format` is used for version comparisons in some places, so just adding `build` to the `format` breaks few tests right away.

**Comment by justinmchase** [comment]
I'm sure you're aware of it but semver.org [says](https://semver.org/#spec-item-10):

> Build metadata MUST be ignored when determining version precedence.

Perhaps `.format()` is fine as is but then the `.toString()` implementation is what should retain the build metadata? I could see how changing this would affect some tests perhaps but I think the intent of the tests themselves is likely still the same, meaning their comparisons should still yield the same results regardless of the build metadata present.

Would you prefer this issue be raised there first and then merged here if/after accepted there?

**Comment by justinmchase** [comment]
Here is a related thread:
https://github.com/npm/node-semver/issues/174

It seems like the developers doing the maintaining perhaps don't even understand the spec itself very well.

Here is what I would recommend:

A) Do it the right way here, call it a feature advantage for Deno over node.
B) Keep the parsing as is, keep the format function as is, it has the side effect of setting the internal `this.version` field, which has only the parts of the version which are significant for determining precedence.
C) Alter the `.toString()` function to emit the version plus the build metadata. Or maybe add a new function if you want to preserve backward compatibility.
D) Add a new function perhaps `.metadata(...build: string[])`, which returns a new semver with the new build metadata replaced.

```ts
import { parse } from "https://deno.land/std@0.next.0/semver/mod.ts";
const v0 = parse('1.2.3+4')
console.log(v0.format()) // 1.2.3
console.log(v0.toString()) // 1.2.3+4

const v1 = v0.metadata('5')
console.log(v1.toString()) // 1.2.3+5
```

**Comment by justinmchase** [comment]
@ben-xD Do you care to comment?

**Comment by kamilogorek** [comment]
Here's a quick PR that one can play with just in case - https://github.com/denoland/deno_std/pull/3117
Although `format` itself is not pure function, so maybe an alternative just for printing would be preferable.

**Comment by ben-xD** [comment]
It's nice that users can get their full version string in Deno. 🎉

**Question:** Should the **comparisons** be changed/fixed, rather than making the API (`Semver#format`) return a version without build metadata? It would be confusing/non-obvious that version strings are incomplete (for those who aren't aware of the unusual behavior of `node-semver`).

**Comment by timreichen** [comment]
How about `FormatOptions` to define how it should be formatted, similar to [Intl DateTime Format](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat)?
```ts
v // 1.2.3-prerelease+meta
format(v, { style: "core" }) // 1.2.3
format(v, { style: "prerelease" }) // 1.2.3-prerelease
format(v, { style: "build" }) // 1.2.3+meta
format(v, { style: "full" }) // (default) 1.2.3-prerelease+meta
```

**Comment by justinmchase** [comment]
I do like @timreichen's suggestion, that does look pretty good.

**Comment by timreichen** [comment]
Maybe `release` would be a more appropriate name whan `core`?

**Comment by justinmchase** [comment]
I made a pretty complete PR based on this discussion: https://github.com/denoland/deno_std/pull/3126

feedback welcome!

**Comment by justinmchase** [comment]
The fix is now available in `0.175.0`
